R/watchdog.R

```diff
@@ -0,0 +1,128 @@
+
+# whether or not the user has enabled the renv watchdog in this session
+`_renv_watchdog_enabled` <- NULL
+
+# metadata related to the running watchdog process, if any
+`_renv_watchdog_metadata` <- NULL
+
+renv_watchdog_init <- function() {
+
+  `_renv_watchdog_enabled` <<- renv_watchdog_enabled_impl()
+  if (!`_renv_watchdog_enabled`)
+    return()
+
+  if (!renv_watchdog_running())
+    renv_watchdog_start()
+
+}
+
+renv_watchdog_enabled <- function() {
+  `_renv_watchdog_enabled`
+}
+
+renv_watchdog_enabled_impl <- function() {
+
+  # skip in older versions of R; we require newer APIs
+  if (getRversion() < "4.0.0")
+    return(FALSE)
+
+  # skip if disabled
+  enabled <- Sys.getenv("RENV_WATCHDOG_ENABLED", unset = "TRUE")
+  if (!truthy(enabled))
+    return(FALSE)
+
+  # skip during R CMD check (but not when running tests)
+  checking <- renv_envvar_exists("_R_CHECK_PACKAGE_NAME_")
+  if (checking && !is_testing())
+    return(FALSE)
+
+  # skip during R CMD build or R CMD INSTALL
+  # ... unless we are running tests on CI
+  building <-
+    renv_envvar_exists("R_PACKAGE_NAME") ||
+    renv_envvar_exists("R_PACKAGE_DIR")
+
+  if (building) {
+    ci <- Sys.getenv("CI", unset = "FALSE")
+    if (!truthy(ci))
+      return(FALSE)
+  }
+
+  # ok, we're enabled
+  TRUE
+
+}
+
+renv_watchdog_start <- function() {
+
+  # create socket for reading initialization data
+  repeat {
+    port <- sample(49152:65536, size = 1L)
+    socket <- tryCatch(serverSocket(port), error = identity)
+    if (!inherits(socket, "error"))
+      break
+  }
```

Comment:
For safety, I think you should terminate this after some number of tries, because if `serverSocket(port)` systematically errors for some reason, you'll get an infinite loop.

Reply:
Good idea; I'll do that.